torch/distributed/_functional_collectives.py

```diff
@@ -363,26 +361,49 @@ def reduce_scatter_tensor(
         tensor_list = torch.chunk(self, group_size, dim=scatter_dim)
         self = torch.cat(tensor_list)
 
-    tensor = torch._C._nn.reduce_scatter_tensor(self, reduceOp, 0, tag, rankset, group_size)  # type: ignore[attr-defined]
+    tensor = torch.ops.c10d_functional.reduce_scatter_tensor(self, reduceOp, tag, rankset, group_size)  # type: ignore[attr-defined]
     res = _maybe_wrap_tensor(tensor)
     return res
 
+c10_lib = torch.library.Library("c10d_functional", "DEF")
+c10_lib_impl = torch.library.Library("c10d_functional", "IMPL")
 
-c10_lib_cpu = torch.library.Library("aten", "IMPL", "CPU")
-c10_lib_cuda = torch.library.Library("aten", "IMPL", "CUDA")
+# We now register meta kernels to deal with tracing
+def _all_reduce_meta(self, *args):
+    return torch.empty_like(self)
 
-def _register_ops():
-    c10_lib_cpu.impl("all_reduce", _all_reduce)
-    c10_lib_cuda.impl("all_reduce", _all_reduce)
+def _wait_tensor_meta(self, *args):
+    return torch.empty_like(self)
+
+def _all_gather_into_tensor_meta(shard, tag, rankset, group_size):
+    out_size = list(shard.size())
+    out_size[0] *= group_size
+    return shard.new_empty(out_size)
 
-    c10_lib_cpu.impl("wait_tensor", _wait_tensor)
-    c10_lib_cuda.impl("wait_tensor", _wait_tensor)
+def _reduce_scatter_tensor_meta(input, reduce_op, tag, rankset, group_size):
+    out_size = list(input.size())
+    out_size[0] //= group_size
+    return input.new_empty(out_size)
+
+
+def _register_ops():
+    ops_defs = [
+        "all_reduce(Tensor self, str reduceOp, str tag, int[] ranks, int group_size) -> Tensor",
+        "wait_tensor(Tensor self) -> Tensor",
+        "all_gather_into_tensor(Tensor shard, str tag, int[] ranks, int group_size) -> Tensor",
+        "reduce_scatter_tensor(Tensor input, str reduceOp, str tag, int[] ranks, int group_size) -> Tensor",
+    ]
+
+    my_module = sys.modules[__name__]
+    for op_def in ops_defs:
+        op_name = op_def[0:op_def.index('(')]
+        backend_impl = getattr(my_module, f"_{op_name}")
+        meta_impl = getattr(my_module, f"_{op_name}_meta")
+        c10_lib.define(op_def)
+        c10_lib_impl.impl(op_name, backend_impl, "CompositeExplicitAutograd")
```

Comment:
would be nice to comment for future readers why we chose to register to CompsiteExplicit

(the core assumption being that both in eager and compiled, we expect these ops to work across all backends who either support processgroup API (eager) or lower the op (compiled)